usaspending_api/common/threaded_data_loader.py

```diff
@@ -0,0 +1,194 @@
+    #                           skip - Skip the row
+    #                           skip_and_complain - Log a warning and skip the row
+    #                           die - Raise an exception and cease execution
+    #   post_row_function - A function to call after every row. Should support parameters of row and instance
+    #                       i.e. def do_this_after_a_row(row=None, instance=None)
+    #   pre_row_function - Like post_row_function, but before the model class is updated
+    #   post_process_function - A function to call when all rows have been processed, uses the same
+    #                           function parameters as post_row_function
+    def __init__(self, model_class, threads=10, field_map={}, value_map={}, collision_field=None, collision_behavior='delete', pre_row_function=None, post_row_function=None, post_process_function=None):
+        self.model_class = model_class
+        self.threads = threads
+        self.field_map = field_map
+        self.value_map = value_map
+        self.collision_field = collision_field
+        self.collision_behavior = collision_behavior
+        self.pre_row_function = pre_row_function
+        self.post_row_function = post_row_function
+        self.post_process_function = post_process_function
+        self.fields = [field.name for field in self.model_class._meta.get_fields()]
+
+    # Loads data from a file using parameters set during creation of the loader
+    # The filepath parameter should be the string location of the file for use with open()
+    def load_from_file(self, filepath):
+        start_time = time.time()
+        print('Started processing file ' + filepath + ' at ' + str(start_time))
+        # Create the Queue object - this will hold all the rows in the CSV
+        row_queue = queue.Queue(maxsize=100)
+        self.exit_flag = False
+
+        # Spawn our threads
+        pool = []
+        while len(pool) < self.threads:
+            thread = DataLoaderThread("Thread-" + str(len(pool)), row_queue, self)
+            thread.start()
+            pool.append(thread)
+
+        with open(filepath) as csvfile:
+            reader = csv.DictReader(csvfile)
+            for row in reader:
+                row_queue.put(row)
+
+        while not row_queue.empty():
+            pass
+
+        self.exit_flag = True
+
+        for thread in pool:
+            thread.join()
+
+        if self.post_process_function is not None:
+            self.post_process_function()
+
+        end_time = time.time()
+        elapsed = end_time - start_time
+        print('Finished processing at ' + str(end_time) + '; Elapsed: ' + str(elapsed))
+
+
+class DataLoaderThread(threading.Thread):
+    def __init__(self, name, data_queue, loader):
+        threading.Thread.__init__(self)
+        self.name = name
+        self.data_queue = data_queue
+        self.loader = loader
+
+    def run(self):
```

Comment:
Yup! It's a pretty common set-up across most languages. Classes implementing threading usually have a run() function which is executed upon the thread object being .start()ed